From fd6ffdb7073bd604e1723f9086f590febae5d7e4 Mon Sep 17 00:00:00 2001
From: Sawan Roy <sawan.roy@trunexa.com>
Date: Mon, 11 Apr 2022 16:18:42 +0530
Subject: [PATCH 5/5] imx8m: Fix the suspend/resume hang issue on imx8mq b0  It
 seems the DRAM APB clock root slice can NOT work normally if the PLLs is 
 power down in DSM mode. So update this clock slice's setting explictly to 
 make it work.

---
 plat/imx/imx8m/ddr/dram_retention.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/plat/imx/imx8m/ddr/dram_retention.c b/plat/imx/imx8m/ddr/dram_retention.c
index 2dd79b6..35c95e9 100644
--- a/plat/imx/imx8m/ddr/dram_retention.c
+++ b/plat/imx/imx8m/ddr/dram_retention.c
@@ -126,6 +126,11 @@ void dram_exit_retention(void)
 	mmio_write_32(CCM_CCGR(5), 2);
 	mmio_write_32(CCM_SRC_CTRL(15), 2);
 
+	/* change the clock source of dram_apb_clk_root */
+	mmio_write_32(0x3038a088, (0x7 << 24) | (0x7 << 16));
+	mmio_write_32(0x3038a084, (0x4 << 24) | (0x3 << 16));
+
+
 	/* disable iso */
 	mmio_setbits_32(IMX_GPC_BASE + PU_PGC_UP_TRG, DDRMIX_PWR_REQ);
 	mmio_write_32(SRC_DDR1_RCR, 0x8F000006);
-- 
2.7.4

